### DISCUSSION TITLE
Predicate harmonisation

### DISCUSSION THREAD

**Discussion by paulakeen** [eprocurementontology/eprocurementontology#517] [Ideas]
Generic predicates are being used frequently in the current ePO linking either multiple domains and multiple ranges; e.g., _epo-ecat:eCatalogue epo:isComposedOf epo-ecat:eCatalogueLine_, but also _epo-eorder:eOrder epo:isComposedOf epo-eorder:eOrderLine_; the property 'epo:hasName' is another typical example in ePO.

This can be a problem when creating the OWL artefacts because the distinction between domains and ranges cannot be differentiated and therefore it could cause inconsistent semantics.

A solution for this could be to document that these properties are so generic that they apply in fact to any domain and any range represented by the owl:Thing, and at the transformation time the semantics can be refined into the specific domain and property (e.g. via restrictions, amongst other).

**Comment by giorgialodi** [comment]
In doing this we may pay attention to the semantic inconsistencies that have been detected by the reasoners. This may entail to change the model.

**Comment by VladimirAlexiev** [comment]
A better approach is to avoid using such generic props: what is the benefit of having them? Would you ever want a query result that mixes such heterogeneous objects?

See related discussion in https://github.com/OP-TED/model2owl/issues/91 (I see @giorgialodi also commented there).

IMHO these are better prop names:
```
ecat:eCatalogue epo:catalogueLine epo-ecat:eCatalogueLine.
epo-eorder:eOrder epo:orderLine epo-eorder:eOrderLine.
```

**Comment by giorgialodi** [comment]
Hi @VladimirAlexiev I don't fully agree with the sentence "A better approach is to avoid using such generic props". We noticed some problems on the use of these properties (for instance "has" and others) that however have been fully revised in the new version of the ontology that is not currently published. Anyway, generic props can be used in specific cases in the modelling, IMHO, especially if you organise your ontologies with an upper level. In the case of e-procurement ontology there is not a proper upper level but I am not so against on the use of generic properties provided that the domains and/or ranges are rather generic too (owl:Thing). For instance, if I observe a specific pattern that applies to many classes, I can use a generic name of the property that identifies the pattern that applies to owl:Thing. The real issue here is how to express  at best these things in UML that is then automatically transformed in OWL. There are some challenges in this process. 

Finally, there are conventions on the naming of the properties (e.g., use of a verb in object properties).
My2cents

**Comment by VladimirAlexiev** [comment]
@giorgialodi I agree with you that there are no absolute rules, and a good judgement is needed.
- eg "name" (or "hasName") is a perfect prop name because it's unambiguous (everyone understands what it is) even though it may appear as high as owl:Thing
- "has" is a bad name because it doesn't describe what is the object to be had.

In my experience, the best convention is that props should be named after their target (range): this gives you specific props yet reusable enough (because they can be applied to any relevant domain).

**Comment by giorgialodi** [comment]
@VladimirAlexiev I totally agree with you on "has" :) And that why is no longer available! Indeed in the revision we tried to apply what you are saying: it is a quite standard practice in the ontology development, AFAIK. 
I provide you an example of what I refer to when I speak of general property. There are many things in the domain that are subject to terms (of various type). One pattern that we observed is that we may have: owl:Thing subjectTo Term ( this latter class  can be then specialized  in subclasses for each type of term). I think this is a general property but it is quite clear. We may call it isSubjectToTerm sure, and we may consider this definitely, but at least in this pattern we know that many things are subject to term (and this is a general statement) and through restrictions we may decide to which term a specific thing is subject to.

**Comment by guascce** [comment]
We will consider this comment when dealing with new properties. For now we have revised the existing ones that were causing issues.
 We do not exclude to keep generic properties when relevant.

**Comment by VladimirAlexiev** [comment]
@giorgialodi hasTerms or isSubjectToTerm is a better prop than isSubjectTo, which invokes ambiguity.

**Comment by eprocurementontology** [comment]
We will take this into consideration when updating the ontology
